File: mobile_app/lib/models/common/service_response.dart
class ServiceResponse<T> {
  final bool isSuccess;
  final T? data;
  final String message;
  final List<String>? errors;

  ServiceResponse({
    required this.isSuccess,
    this.data,
    required this.message,
    this.errors,
  });

  factory ServiceResponse.fromJson(
    Map<String, dynamic> json,
    T Function(dynamic)? dataDecoder,
  ) {
    return ServiceResponse(
      isSuccess: json['isSuccess'] ?? false,
      data: json['data'] != null ? dataDecoder?.call(json['data']) : null,
      message: json['message'] ?? '',
      errors: json['errors'] != null ? List<String>.from(json['errors']) : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'isSuccess': isSuccess,
        'data': data,
        'message': message,
        'errors': errors,
      };
}

File: backend/Common/Results/ServiceResponse.cs
namespace DataLabellingSupportSystem.Api.Common.Results;

public class ServiceResponse<T>
{
    public bool IsSuccess { get; set; }
    public T? Data { get; set; }
    public string Message { get; set; } = string.Empty;
    public List<string> Errors { get; set; } = [];

    public static ServiceResponse<T> Success(T data, string message = "Success")
    {
        return new ServiceResponse<T>
        {
            IsSuccess = true,
            Data = data,
            Message = message
        };
    }

    public static ServiceResponse<T> Failure(string message, List<string>? errors = null)
    {
        return new ServiceResponse<T>
        {
            IsSuccess = false,
            Message = message,
            Errors = errors ?? []
        };
    }
}
